library/system/src/modules/Kernel.rb

```diff
@@ -731,30 +670,93 @@ def read_modules_to_load
-    publish function: :ComputePackage, type: "string ()"
-    publish function: :GetFinalKernel, type: "string ()"
-    publish function: :ComputePackagesForBase, type: "list <string> (string, boolean)"
-    publish function: :ComputePackages, type: "list <string> ()"
-    publish function: :SetInformAboutKernelChange, type: "void (boolean)"
-    publish function: :GetInformAboutKernelChange, type: "boolean ()"
-    publish function: :InformAboutKernelChange, type: "boolean ()"
+    # @param [String] line to parse
+    # @return [Array<String>] line splitted to individual params, respecting quotes there
+    def list_of_params(line)
+      line = line.delete("\n")
+      cmdlist = []
+      parse_index = 0
+      in_quotes = false
+      after_backslash = false
+      current_param = ""
+      while parse_index < line.size
+        current_char = line[parse_index]
+        in_quotes = !in_quotes if current_char == "\"" && !after_backslash
+        if current_char == " " && !in_quotes
+          cmdlist << current_param
+          current_param = ""
+        else
+          current_param << current_char
+        end
+        # does kernel cmdline can have double backslash?
+        after_backslash = current_char == "\\"
+        parse_index += 1
+      end
+      cmdlist << current_param
+    end
 
-    # Handling for Kernel modules loaded on boot
-    publish function: :AddModuleToLoad, type: "void (string)"
-    publish function: :RemoveModuleToLoad, type: "void (string)"
-    publish function: :SaveModulesToLoad, type: "boolean ()"
-    publish function: :reset_modules_to_load, type: "void ()"
-    publish function: :modules_to_load, type: "map <string, list> ()"
+    # constructs list of keys to discard from command line
+    # @param [Array<String>] list of command line entries
+    # @return [Array<String>] list of keys to discard
+    def params_to_discard(cmdlist)
+      discardlist = []
+      # some systems (pseries) can autodetect the serial console
+      if cmdlist.include?("AUTOCONSOLE")
+        discardlist << "console"
+        discardlist << "AUTOCONSOLE"
+      end
+
+      # add special key filtering for s390
+      # bnc#462276 Extraneous parameters in /etc/zipl.conf from the installer
+      discardlist << "User" << "init" << "ramdisk_size" if Arch.s390
+
+      # get rid of live-installer-specific parameters
+      if Mode.live_installation
+        discardlist.push("initrd", "ramdisk_size", "ramdisk_blocksize", "liveinstall", "splash", "quiet", "lang")
+      end
+
+      # TODO: is it still needed?
+      # backdoor to re-enable update on UL/SLES
+      if cmdlist.include?("suse_update")
+        discardlist << "suse_update"
+        @suse_update = true
+      end
+
+      discardlist
+    end
+
+    # @param	[String] cmdline to parse
+    #
+    # @return	[void]
+    # Filters out yast2 specific boot parameters and sets
+    # Parameters to the important cmdline parts.
+    def ExtractCmdlineParameters(line)
```

Comment:
As this is (not yet, but it should) a private method too, I would much prefer a cleaner calling sequence. But it is a NTH, an encapsulated implementation detail.

```rb
def ParseInstallationKernelCmdline
  cmd_string = if Mode... ...; end
  log.info ... HidePasswords(cmd_string)
  cmd_list = tokenize(cmd_string)            # renamed list_of_params
  to_reject = params_to_discard(cmd_list)
  cmd_list.reject! { |c| params_to_discard.include?(c) }
  assign_params(cmd_list)                    # renamed rest of ExtractCmdlineParameters
end